lib/babe/build.go

```diff
@@ -18,25 +18,79 @@ package babe
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
 	"math/big"
 	"time"
 
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+
 	"github.com/ChainSafe/gossamer/dot/types"
 	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/runtime"
 	"github.com/ChainSafe/gossamer/lib/scale"
 	"github.com/ChainSafe/gossamer/lib/transaction"
 )
 
-// BuildBlock builds a block for the slot with the given parent.
-// TODO: separate block builder logic into separate module. The only reason this is exported is so other packages
-// can build blocks for testing, but it would be preferred to have the builder functionality separated.
-func (b *Service) BuildBlock(parent *types.Header, slot Slot) (*types.Block, error) {
-	return b.buildBlock(parent, slot)
-}
-
 // construct a block for this slot with the given parent
 func (b *Service) buildBlock(parent *types.Header, slot Slot) (*types.Block, error) {
+	builder, err := NewBuilder(
+		b.rt,
+		b.keypair,
+		b.transactionState,
+		b.blockState,
+		b.slotToProof,
+		b.epochData.authorityIndex,
+	)
+
+	if err != nil {
+		return nil, errors.New("There was an error creating block builder - " + err.Error())
+	}
+
+	block, err := builder.buildBlock(parent, slot)
+
+	return block, err
+
+}
+
+// nolint
+type BlockBuilder struct {
+	rt                    runtime.Instance
+	keypair               *sr25519.Keypair
+	transactionState      TransactionState
+	blockState            BlockState
+	slotToProof           map[uint64]*VrfOutputAndProof
+	currentAuthorityIndex uint32
+}
+
+// nolint
+func NewBuilder(rt runtime.Instance, kp *sr25519.Keypair, ts TransactionState, bs BlockState, sp map[uint64]*VrfOutputAndProof, authidx uint32) (*BlockBuilder, error) {
```

Comment:
The name could be `NewBlockBuilder` as it is related to construct a `BlockBuilder`

Reply:
Done!